keras/caffe/converters.py

```diff
@@ -0,0 +1,470 @@
+    network = remove_label_paths(network, network_inputs, network_outputs)  # path from input to loss layers removed
+    reverse_network = reverse(network)  # stores the 'input' for each layer
+
+    model = Graph()
+
+    # add input nodes
+    for network_input in network_inputs:
+        name = layers[network_input].name
+        model.add_input(name=name, ndim=4)
+
+    # parse all the layers and build equivalent keras graph
+    for layer_nb in network:
+        layer = layers[layer_nb]
+        name = layer.name
+
+        if is_data_input(layer):
+            continue
+
+        if layer_nb in network_inputs:
+            # this is both an input layer and computation layer. the name has been added as input layer
+            input_layer_names = [name]
+            # use a prefix to create a new name, which is used on from hereon to take input from this layer(see below)
+            name = 'graph_input_' + name
+        else:
+            input_layers = reverse_network[layer_nb]  # inputs to current layer, in the form of layer numbers
+            input_layer_names = []  # list of strings identifying the input layer
+            for input_layer in input_layers:
+                # taking input from first layer and that's not a data layer, implies use the prefix
+                if input_layer in network_inputs and not is_data_input(layers[input_layer]):
+                    input_layer_names.append('graph_input_' + layers[input_layer].name)
+                else:
+                    input_layer_names.append(layers[input_layer].name)
+
+        if len(input_layer_names) == 1:
+            # single input. since concatenation is explicit,
+            # all layers can be thought of single input layers
+            # (except loss layers, which is handled anyway)
+            input_layer_name = input_layer_names[0]
+
+        if layer_nb in network_outputs:
+            # outputs nodes are marked with 'output_' prefix from which output is derived
+            # since the graph output is not model output
+            name = 'output_' + name
+
+        if layer.type == 3:
+            # CONCAT
+            # emulation of just concatenation
+            axis = layer.concat_param.axis  # 0 for batch, 1 for stack
+            model.add_node(Activation('linear'), name=name, inputs=input_layer_names, concat_axis=axis)
+
+        elif layer.type == 4:
+            # CONVOLUTION
+            blobs = layer.blobs
+            nb_filter, temp_stack_size, nb_col, nb_row = blobs[0].num, blobs[0].channels, blobs[0].height, blobs[0].width
+
+            # weird old trick used in paralellizing networks
+            group = layer.convolution_param.group
+            stack_size = temp_stack_size * group
+
+            # maybe not all synapses are existant
+            weights_p = np.zeros((nb_filter, stack_size, nb_col, nb_row))
+            weights_b = np.array(blobs[1].data)
+
+            chunk_data_size = len(blobs[0].data) // group
```

Comment:
Just for reference, this is how chainer does this:
```python
def _setup_convolution(self, layer):
        blobs = layer.blobs
        param = layer.convolution_param
        ksize = _get_ksize(param)
        stride = _get_stride(param)
        pad = _get_pad(param)
        num = _get_num(blobs[0])
        channels = _get_channels(blobs[0])

        n_in = channels * param.group
        n_out = num
        func = functions.Convolution2D(n_in, n_out, ksize, stride, pad,
                                       nobias=not param.bias_term)
        func.W.fill(0)

        part_size = len(blobs[0].data) // param.group
        for i in six.moves.range(param.group):
            in_slice = slice(i * n_in // param.group,
                             (i+1) * n_in // param.group)
            out_slice = slice(i * n_out // param.group,
                              (i+1) * n_out // param.group)
            w = func.W[out_slice, in_slice]

            data = numpy.array(blobs[0].data[i*part_size:(i+1)*part_size])
            w[:] = data.reshape(w.shape)

        if param.bias_term:
            func.b[:] = blobs[1].data

        setattr(self.fs, layer.name, func)
        self.forwards[layer.name] = func
        self._add_layer(layer)
```

And this is how sklearn-theano does this:
def _blob_to_ndarray(blob):
```python
    """Converts a caffe protobuf blob into an ndarray"""
    dimnames = ["num", "channels", "height", "width"]
    data = np.array(blob.data)
    shape = tuple([getattr(blob, dimname) for dimname in dimnames])
    return data.reshape(shape)
```